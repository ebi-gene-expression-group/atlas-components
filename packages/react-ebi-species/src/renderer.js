"use strict";

//*------------------------------------------------------------------*

var React = require('react');
var ReactDOM = require('react-dom');

//*------------------------------------------------------------------*

var Icon = require('./SpeciesIcon.jsx');

//*------------------------------------------------------------------*

module.exports = function(mountNode,species,colourOverride, colourPerKingdom) {
    ReactDOM.render(
        React.createElement(Icon,{
          species: species,
          colourOverride: colourOverride,
          colourPerKingdom: colourPerKingdom
        }), mountNode
    );
};
