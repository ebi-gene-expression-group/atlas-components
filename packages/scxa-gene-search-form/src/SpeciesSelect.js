import React from 'react'
import PropTypes from 'prop-types'
import Select from 'react-select'

import ebiVfReactSelectReplacements from './ebiVfReactSelectReplacements'

const _makeOption = (str) => {
  if (str === ``) {
    return {
      value: ``,
      label: `Any`
    }
  }

  return {
    value: str.trim(),
    label: str.trim()
  }
}

const _onlyUnique = (value, index, self) => self.indexOf(value) === index

const SpeciesSelect = ({topSpecies, allSpecies, statusMessage, onChange, selectedSpecies}) => {
  const topOptions = topSpecies.map(_makeOption)

  const allOptions =
    allSpecies
      .concat(``)
      .map(str => str.trim())
      .filter(_onlyUnique)
      .sort()
      .map(_makeOption)

  const options =
    topOptions
      .concat([{
        label: Math.random() < 0.9999 ? `All species` : `(╯°□°）╯︵ ┻━┻`,
        options: allOptions
      }])

  const selectedValue =
    topOptions
      .concat(allOptions)
      .find(option => option.value === selectedSpecies.trim()) || allOptions[0]

  return (
    <div>
      <label  key={`label`} htmlFor={`species`}>Species</label>
      <Select
        key={`select`}
        name={`species`}
        components={{ IndicatorSeparator: null, DropdownIndicator: ebiVfReactSelectReplacements.DropdownIndicator }}
        styles={ebiVfReactSelectReplacements.styles}
        isSearchable={false}
        onChange={onChange}
        options={options}
        isDisabled={Boolean(statusMessage)}
        defaultValue={Boolean(statusMessage) ? null : selectedValue}
        placeholder={Boolean(statusMessage) ? statusMessage : null} />
    </div>
  )
}

SpeciesSelect.propTypes = {
  topSpecies: PropTypes.arrayOf(PropTypes.string),
  allSpecies: PropTypes.arrayOf(PropTypes.string),
  statusMessage: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  selectedSpecies: PropTypes.string
}

SpeciesSelect.defaultProps = {
  topSpecies: [],
  allSpecies: [],
  statusMessage: ``,
  selectedSpecies: ``
}

export default SpeciesSelect
