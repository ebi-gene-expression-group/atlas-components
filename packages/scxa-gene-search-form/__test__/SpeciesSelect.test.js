import React from 'react'
import renderer from 'react-test-renderer'
import Enzyme from 'enzyme'
import { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import Select from 'react-select'
import SpeciesSelect from '../src/SpeciesSelect'

Enzyme.configure({ adapter: new Adapter() })

const getRandomInt = (max) => Math.floor(Math.random() * Math.floor(max))

const props = {
  topSpecies: [
    `Meeseek`,
    `Zigerion`
  ],
  allSpecies: [
    `Meeseek`,
    `Gromflomite`,
    `Cromulon`,
    `Zigerion`,
    `Moopian`,
    `Bliznarvian`,
    `Greebybobe`
  ],
  statusMessage: null,
  onChange: ()=> {},
  selectedSpecies: ``
}

describe(`SpeciesSelect`, () => {
  test(`is made up of a label and a React Select component`, () => {
    const wrapper = shallow(<SpeciesSelect {...props} />)
    expect(wrapper.first().find(`label`)).toHaveLength(1)
    expect(wrapper.first().find(`label`).text()).toEqual(`Species`)
    expect(wrapper.last().find(Select)).toHaveLength(1)
  })

  test(`is disabled when a statusMessage (i.e. an error message) is passed in and it is displayed`, () => {
    const errorMessage = `Any schwifty, non-empty string`
    const wrapper = shallow(<SpeciesSelect {...props} statusMessage={errorMessage} />)
    expect(wrapper.find(Select).prop(`isDisabled`)).toEqual(true)
    expect(wrapper.find(Select).prop(`placeholder`)).toEqual(errorMessage)
  })

  test(`is enabled when statusMessage is null`, () => {
    const wrapper = shallow(<SpeciesSelect {...props} />)
    expect(wrapper.find(Select).prop(`isDisabled`)).toEqual(false)
  })

  test(`selects Any species by default`, () => {
    const wrapper = shallow(<SpeciesSelect {...props} />)
    expect(wrapper.find(Select).prop(`defaultValue`)).toHaveProperty(`value`, ``)
    expect(wrapper.find(Select).prop(`defaultValue`)).toHaveProperty(`label`, `Any`)
  })

  test(`preselects the selectedSpecies prop`, () => {
    const selectedSpecies = props.allSpecies[getRandomInt(props.allSpecies.length)]
    const wrapper = shallow(<SpeciesSelect {...props} selectedSpecies={selectedSpecies}/>)
    expect(wrapper.find(Select).prop(`defaultValue`)).toHaveProperty(`value`, selectedSpecies)
  })

  test(`selects Any if selecetdSpecies doesnâ€™t match any in topSpecies or allSpecies`, () => {
    const wrapper = shallow(<SpeciesSelect {...props} selectedSpecies={`Human`}/>)
    expect(wrapper.find(Select).prop(`defaultValue`)).toHaveProperty(`value`, ``)
    expect(wrapper.find(Select).prop(`defaultValue`)).toHaveProperty(`label`, `Any`)
  })

  test(`top species may not be included in all species`, () => {
    const topSpecies = [`Human`]
    const wrapper = shallow(<SpeciesSelect {...props} topSpecies={topSpecies}/>)

    // allSpecies group is the one which has an options property instead of a value one
    const allSpeciesGroup = wrapper.find(Select).prop(`options`).find(obj => Boolean(obj.options))

    expect(allSpeciesGroup.options).toContainEqual({
      value: props.allSpecies[getRandomInt(props.allSpecies.length)],
      label: expect.stringMatching(/.*/)
    })

    expect(allSpeciesGroup.options).not.toContainEqual({
      value: topSpecies[0],
      label: expect.stringMatching(/.*/)
    })
  })

  test(`changing the value in the React Select component fires the onChange prop function`, () => {
    const onChangeMock = jest.fn()
    const wrapper = shallow(<SpeciesSelect {...props} onChange={onChangeMock} />)

    const speciesToSelect = props.allSpecies[getRandomInt(props.allSpecies.length)]
    wrapper.find(Select).simulate(`change`, { value: speciesToSelect, label: speciesToSelect })

    expect(onChangeMock.mock.calls).toHaveLength(1)
    expect(onChangeMock.mock.calls[0][0]).toEqual({ value: speciesToSelect, label: speciesToSelect });
  })

  test(`matches snapshot`, () => {
    const tree = renderer.create(<SpeciesSelect {...props}/>).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
