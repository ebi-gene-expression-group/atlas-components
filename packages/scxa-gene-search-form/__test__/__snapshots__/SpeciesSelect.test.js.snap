// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpeciesSelect matches snapshot 1`] = `
<div>
  <label
    htmlFor="species"
  >
    Species
  </label>
  <div
    className="css-10nd86i"
    onKeyDown={[Function]}
  >
    <div
      className="css-12emboz"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="css-1hwfws3"
      >
        <div
          className="css-xp4uvy"
        >
          Any
        </div>
        <input
          className="css-14uuagi"
          id="react-select-2-input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          readOnly={true}
          tabIndex="0"
          value=""
        />
      </div>
      <div
        className="css-1wy0on6"
      >
        <span
          className="sc-bdVaJa jcfyVD"
          cx={[Function]}
        />
      </div>
    </div>
    <input
      name="species"
      type="hidden"
      value=""
    />
  </div>
</div>
`;
